src/Util/FileMatcher.php

```diff
@@ -0,0 +1,297 @@
+                $resolved[] = [self::T_CHAR, $char];
+
+                continue;
+            }
+
+            // https://man7.org/linux/man-pages/man7/glob.7.html
+            // > The string enclosed by the brackets cannot be empty; therefore
+            // > ']' can be allowed between the brackets, provided that it is
+            // > the first character.
+            if ($type === self::T_BRACKET_OPEN && $nextType === self::T_BRACKET_CLOSE) {
+                $bracketOpen = true;
+                $resolved[]  = [self::T_BRACKET_OPEN, '['];
+                $brackets[]  = array_key_last($resolved);
+                $resolved[]  = [self::T_CHAR, ']'];
+                $offset++;
+
+                continue;
+            }
+
+            // if we're already in a bracket and the next two chars are [: then
+            // start parsing a character class...
+            if ($bracketOpen && $type === self::T_BRACKET_OPEN && $nextType === self::T_COLON) {
+                // this looks like a named [:character:] class
+                $class = '';
+                $offset += 2;
+
+                // parse the character class name
+                while (ctype_alpha($tokens[$offset][1])) {
+                    $class .= $tokens[$offset++][1];
+                }
+
+                // if followed by a `:` then it's a character class
+                if ($tokens[$offset][0] === self::T_COLON) {
+                    $offset++;
+                    $resolved[] = [self::T_CHAR_CLASS, $class];
+
+                    continue;
+                }
+
+                // otherwise it's a harmless literal
+                $resolved[] = [self::T_CHAR, ':' . $class];
+            }
+
+            // if bracket is already open and we have another open bracket
+            // interpret it as a literal
+            if ($bracketOpen === true && $type === self::T_BRACKET_OPEN) {
+                $resolved[] = [self::T_CHAR, $char];
+
+                continue;
+            }
+
+            // if we are NOT in an open bracket and we have an open bracket
+            // then pop the bracket on the stack and enter bracket-mode.
+            if ($bracketOpen === false && $type === self::T_BRACKET_OPEN) {
+                $bracketOpen = true;
+                $resolved[]  = [$type, $char];
+                $brackets[]  = array_key_last($resolved);
+
+                continue;
+            }
+
+            // if are in a bracket and we get to bracket close then
+            // pop the last open bracket off the stack and continue
+            //
+            // TODO: $bracketOpen === true below is not tested
+            if ($bracketOpen === true && $type === self::T_BRACKET_CLOSE) {
+                // TODO: this is not tested
+                $bracketOpen = false;
+
+                array_pop($brackets);
+                $resolved[] = [$type, $char];
+
+                continue;
+            }
+
+            $resolved[] = [$type, $char];
```

Comment:
this handles any "left over tokens" - including `T_ASTERISK` - maybe all tokens should be handled explicitly?